File: backend/src/main/java/com/mini/credit/service/EmployeService.java
package com.mini.credit.service;

import com.mini.credit.dto.employe.EmployeDTO;
import com.mini.credit.dto.employe.CreateEmployeRequest;
import com.mini.credit.dto.employe.UpdateEmployeRequest;

import java.util.List;
import org.springframework.web.multipart.MultipartFile;

public interface EmployeService {
    EmployeDTO create(CreateEmployeRequest request);
    EmployeDTO update(Long id, UpdateEmployeRequest request);
    EmployeDTO getById(Long id);
    EmployeDTO getByMatricule(String matricule);
    List<EmployeDTO> getAll();
    List<EmployeDTO> getAllActive();
    /** Employés actifs sans compte utilisateur lié — pour le dropdown "Créer utilisateur" */
    List<EmployeDTO> getDisponibles();
    /** Employés actifs avec fonction = GESTIONNAIRE — pour le dropdown "Créer Agent Terrain" */
    List<EmployeDTO> getGestionnaires();
    /** Employés actifs rattachés à une agence donnée — pour le détail Agence */
    List<EmployeDTO> getEmployesByAgence(Long agenceId);
    /**
     * Employés actifs n'appartenant PAS à l'agence donnée — candidats au transfert.
     * Utilisé par la modal "Affecter un employé existant" dans le détail Agence.
     */
    List<EmployeDTO> getEmployesAAffecterAgence(Long agenceId);
    /**
     * Transfère un employé vers une nouvelle agence (et un site de cette agence).
     * Le site doit appartenir à l'agence cible.
     * Réservé à l'ADMIN.
     */
    EmployeDTO changerAgence(Long employeId, Long agenceId, Long siteId);
    EmployeDTO uploadPhoto(Long employeId, MultipartFile file);
    EmployeDTO deletePhoto(Long employeId);
    void delete(Long id);
}

File: frontend/src/app/features/employes/services/employe.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { API_BASE_URL } from '../../../core/services/api.config';
import { EmployeResponse } from '../models/employe-response';
import { CreateEmployeRequest } from '../models/employe-create-request';
import { UpdateEmployeRequest } from '../models/employe-update-request';

@Injectable({
  providedIn: 'root'
})
export class EmployeService {
  private readonly apiUrl = `${API_BASE_URL}/employes`;

  constructor(private http: HttpClient) {}

  getAll(): Observable<EmployeResponse[]> {
    return this.http.get<EmployeResponse[]>(this.apiUrl);
  }

  getAllActive(): Observable<EmployeResponse[]> {
    return this.http.get<EmployeResponse[]>(`${this.apiUrl}/actifs`);
  }

  /**
   * Employés actifs sans compte utilisateur lié.
   * Utilisé par le formulaire "Créer Utilisateur" pour le dropdown.
   */
  getDisponibles(): Observable<EmployeResponse[]> {
    return this.http.get<EmployeResponse[]>(`${this.apiUrl}/disponibles`);
  }

  getById(id: number): Observable<EmployeResponse> {
    return this.http.get<EmployeResponse>(`${this.apiUrl}/${id}`);
  }

  getByMatricule(matricule: string): Observable<EmployeResponse> {
    return this.http.get<EmployeResponse>(`${this.apiUrl}/matricule/${matricule}`);
  }

  create(request: CreateEmployeRequest): Observable<EmployeResponse> {
    return this.http.post<EmployeResponse>(this.apiUrl, request);
  }

  update(id: number, request: UpdateEmployeRequest): Observable<EmployeResponse> {
    return this.http.put<EmployeResponse>(`${this.apiUrl}/${id}`, request);
  }

  uploadPhoto(id: number, file: File): Observable<EmployeResponse> {
    const formData = new FormData();
    formData.append('file', file);
    return this.http.post<EmployeResponse>(`${this.apiUrl}/${id}/photo`, formData);
  }

  deletePhoto(id: number): Observable<EmployeResponse> {
    return this.http.delete<EmployeResponse>(`${this.apiUrl}/${id}/photo`);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  /**
   * Employés actifs rattachés à une agence donnée.
   * Utilisé par le détail Agence pour afficher le personnel.
   */
  getByAgence(agenceId: number): Observable<EmployeResponse[]> {
    return this.http.get<EmployeResponse[]>(`${this.apiUrl}/by-agence/${agenceId}`);
  }

  /**
   * Transfère un employé vers une nouvelle agence.
   * Le site doit appartenir à l'agence cible.
   */
  changerAgence(id: number, agenceId: number, siteId: number): Observable<EmployeResponse> {
    return this.http.patch<EmployeResponse>(`${this.apiUrl}/${id}/agence`, { agenceId, siteId });
  }

  /**
   * Employés actifs n'appartenant PAS à l'agence donnée.
   * Utilisé par la modal "Affecter un employé existant".
   */
  getEmployesAAffecterAgence(agenceId: number): Observable<EmployeResponse[]> {
    return this.http.get<EmployeResponse[]>(`${this.apiUrl}/a-affecter-agence/${agenceId}`);
  }
}
